### PR TITLE
Decommit memory before returning to the region_allocator

### PR DESCRIPTION
**Pull request by cshung**
When we `decommit_region`, we wanted to also de-commit the associated mark array. The method `decommit_mark_array_by_seg` requires a `gc_heap` instance, so I just used the one stored on the region.

For regions that were returned to the free list earlier, their heaps were reset to `null`, so we will be using `null` to de-commit the mark array, and that will fail.

Instead of figuring out what heap it was. It is easier to recognize that in the case of `USE_REGIONS`, the `mark_array` is always the same across all heaps, and the partially committed mark array case just won't happen with `USE_REGIONS`. Therefore, we can simply pick an arbitrary heap and use it to de-commit the mark array.

### CODE REVIEW

**Review by sebastienros — APPROVED** [review]

File: src/coreclr/gc/gc.cpp
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:41131]: this is the case with segments too as the mark array is the same for all heaps. I think it would be more useful to say in the comment why you would hit this with regions (because we clear the heap for free regions).

**Comment by eerhardt** [comment]
Is there any way to add tests for this? So that we don't regress in this area again. It took a bit of time just to diagnose this issue in ASP.NET's CI.

**Comment by sebastienros** [comment]
Looks like this was a "revert PR" that was later changed to a fix. I would have preferred the previous PR to be merged as-is to unblock the aspnet CI and the benchmarks while the fix is being worked on.

**Comment by sebastienros** [comment]
https://github.com/dotnet/runtime/pull/80640#issuecomment-1383253372

**Comment by kasperk81** [comment]
@cshung or @jkotas can you merge this pr to unblock https://github.com/dotnet/aspnetcore/pull/46055. aspnetcore is blocking sdk and installer

**Comment by jkotas** [comment]
I won't merge this without a sign-off from @Maoni0 or somebody else from the GC team.

I can merge revert of the PR that introduced the break. I agree with @sebastienros that it should have been the first course of action: #80723

**Comment by jkotas** [comment]
> I can merge revert of the PR that introduced the break. I agree with @sebastienros that it should have been the first course of action: https://github.com/dotnet/runtime/pull/80723

Done.

**Comment by Maoni0** [comment]
I was waiting on Andrew to address the incorrect comment. I see that @jkotas already reverted the PR. thanks Jan.

**Review by halter73 — APPROVED** [review]

**Review by Maoni0 — APPROVED** [review]

**Comment by eerhardt** [comment]
> Is there any way to add tests for this? So that we don't regress in this area again. It took a bit of time just to diagnose this issue in ASP.NET's CI.

Was there no way to write tests for the regression?

**Comment by cshung** [comment]
> Was there no way to write tests for the regression?

There is no easy way to accomplish that regression testing within the runtime CI, but we have established new processes to make sure GC PRs are validated against the ASP.NET Benchmarks for functional and perf validation. @mrsharm is building automation around that, and this PR is tested using it.

**Pull request merged**
